### PR TITLE
fix(rust, python): Edge cases for list count formatting

### PR DESCRIPTION
**Pull request by Walnut356**
As mentioned [here](https://github.com/pola-rs/polars/pull/11409#issuecomment-1764702316), the values -1, 0, and 1 produce erroneous results. 

The issue with 0 and 1 is that the match statement matches on `self.len()` without considering the config setting. The issue with -1 is that the ellipsis is not optional.

I've restructured the function to handle 4 cases:

1. if the list is empty, always print "[]"

2. if max_items is 0, always print "[…]" for non-empty lists

3. if max_items is greater than or equal to self.len(), print the entire list

4. if max_items is less than self.len(), print numbers up to max_items - 1, print an ellipsis, then print the last item. This means that if max_items is 1, only the last item will be printed.

Worth noting that if max_items is -1, it is set to self.len(), therefore will always be handled by case 2, thus avoiding the ellipsis issue.

I have also updated the docs, and the test now includes the following max_items settings: default value (3), 0, 1, -1, custom value greater than self.len (10), and custom value less than self.len (4). I also added a few special tests for lists of length 1 and empty lists to confirm that those are handled correctly

### CODE REVIEW

**Comment by Walnut356** [comment]
Unfortunately I don't think I know quite enough to figure out why those checks failed. For it being in a spot so far from anything I changed, I'd guess that it has to do with how the list_short and list_empty are created in test_fmt_list, but even if that's the case I'm unsure of how to proceed with fixing it.

**Comment by Walnut356** [comment]
Nevermind, I think I've figured it out. 

* Indexing into the series? Perfect.

* Calling .iter().enumerate()? Totally fine.

* Calling JUST .iter()? Sometimes a disaster for some reason. Iunno.

Again, I'm not expert, so take this with a grain of salt, but I did notice the [printed tables in the test output](https://github.com/pola-rs/polars/actions/runs/6542452699/job/17765505970?pr=11780#step:8:231) were garbled. The last time i saw that, it was because some other part of my program was doing get_unchecked_mut and writing out-of-bounds by ~10-20 thousand bytes (whoops). The program would "complete", but the console output would have messed up borders. 

No idea why this occurs here, but it may be worth looking into ^^

File: crates/polars-core/src/fmt.rs [hunk 1]
```diff
@@ -964,36 +964,46 @@ fn fmt_struct(f: &mut Formatter<'_>, vals: &[AnyValue]) -> fmt::Result {
...
 
-                for (i, item) in self.iter().enumerate() {
+                for i in 0..self.len() {
+                    let item = self.get(i).unwrap();
                     write!(result, "{item}").unwrap();
+                    // this will always leave a trailing ", " after the last item
+                    // but for long lists, this is faster than checking against the length each time
+                    result.push_str(", ");
+                }
+                // remove trailing ", " and replace with closing brace
```
Comment (ritchie46) [inline crates/polars-core/src/fmt.rs:989]: if self.is_empty we shouldn't pop right?

Comment (Walnut356) [reply crates/polars-core/src/fmt.rs:989]: empty lists shouldn't make it past the guard clause on [line 967](https://github.com/pola-rs/polars/pull/11780/files#diff-c794930ce2f11449d29e5f036c31df358a121a713f0ecbed1cfc07a781f695b0R967), so this loop will always run at least once

**Review by ritchie46 — APPROVED** [review]
